Realm/Tests/AsyncTests.m

```diff
@@ -0,0 +1,808 @@
+            [IntObject createInDefaultRealmWithValue:@[@(value)]];
+        }];
+    }
+}
+
+- (void)testInitialResultsAreDelivered {
+    [self createObject:1];
+
+    XCTestExpectation *expectation = [self expectationWithDescription:@""];
+    id token = [[IntObject objectsWhere:@"intCol > 0"] deliverOnMainThread:^(RLMResults *results, NSError *e) {
+        XCTAssertNil(e);
+        XCTAssertEqualObjects(results.objectClassName, @"IntObject");
+        XCTAssertEqual(results.count, 1U);
+        [expectation fulfill];
+    }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+    [token stop];
+}
+
+- (void)testNewResultsAreDeliveredAfterLocalCommit {
+    __block XCTestExpectation *expectation = [self expectationWithDescription:@""];
+    __block NSUInteger expected = 0;
+    id token = [[IntObject objectsWhere:@"intCol > 0"] deliverOnMainThread:^(RLMResults *results, NSError *e) {
+        XCTAssertEqual(results.count, expected++);
+        [expectation fulfill];
+    }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+
+    expectation = [self expectationWithDescription:@""];
+    [self createObject:1];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+
+    expectation = [self expectationWithDescription:@""];
+    [self createObject:2];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+    [token stop];
+}
+
+- (void)testNewResultsAreDeliveredAfterBackgroundCommit {
+    __block XCTestExpectation *expectation = [self expectationWithDescription:@""];
+    __block NSUInteger expected = 0;
+    id token = [[IntObject objectsWhere:@"intCol > 0"] deliverOnMainThread:^(RLMResults *results, NSError *e) {
+        XCTAssertEqual(results.count, expected++);
+        [expectation fulfill];
+    }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+
+    expectation = [self expectationWithDescription:@""];
+    [self dispatchAsyncAndWait:^{ [self createObject:1]; }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+
+    expectation = [self expectationWithDescription:@""];
+    [self dispatchAsyncAndWait:^{ [self createObject:2]; }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+    [token stop];
+}
+
+- (void)testResultsPerserveQuery {
+    __block XCTestExpectation *expectation = [self expectationWithDescription:@""];
+    __block NSUInteger expected = 0;
+    id token = [[IntObject objectsWhere:@"intCol > 0"] deliverOnMainThread:^(RLMResults *results, NSError *e) {
+        XCTAssertEqual(results.count, expected);
+        [expectation fulfill];
+    }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+
+    expectation = [self expectationWithDescription:@""];
+    ++expected;
+    [self dispatchAsyncAndWait:^{ [self createObject:1]; }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+
+    expectation = [self expectationWithDescription:@""];
+    [self dispatchAsyncAndWait:^{ [self createObject:-1]; }];
```

Comment:
What kinds of changes will or will not trigger a redelivery? Will writing to a different column of an observed table trigger one?